feat: 회원가입

smh0300 opened the pull request:
#8
- [x] 회원가입 MVC 구현
- [x] 회원가입 공통 예외처리 적용
- [x] JPA 연동
- [x] 많은 테스트코드 작성
## To Reviewers
- MVC간 분리가 적절한지 검토 부탁드립니다.
- 테스트코드 작성이 적절한지 검토 부탁드립니다.

smh0300 committed:
feat: MainController 작성

smh0300 committed:
Revert "feat: 회원가입(signup) 컨트롤러 구현"

This reverts commit d942d7299a7bb8b4d69695f50387aae443e7063f.

smh0300 committed:
feat: 회원가입(signup) 컨트롤러 구현

smh0300 committed:
feat: signup 뷰 및 컨트롤러 구현

smh0300 committed:
feat: 회원가입 구현

smh0300 committed:
bug: test실패 조치

smh0300 committed:
Update autotest.yml

smh0300 committed:
bug: github action test failure

smh0300 committed:
Merge branch 'feat/signup' of https://github.com/f-lab-edu/TicketEase into feat/signup

smh0300 committed:
fix: springboot datasource error fix

smh0300 committed:
fix: springboot datasource error

smh0300 committed:
fix: springboot datasource error

smh0300 committed:
fix: springboot datasource error

smh0300 committed:
bug: springboot datasource error

smh0300 committed:
fix: springboot datasource error

smh0300 committed:
fix: springboot datasource error

smh0300 committed:
fix: springboot datasource error

smh0300 committed:
fix: springboot datasource error

smh0300 commented on src/main/resources/application-prod.yml:
```diff
@@ -0,0 +1,17 @@
+        dialect: org.hibernate.dialect.MySQLDialect
+    defer-datasource-initialization: 'true'
+    hibernate:
+      # create 안쓰려면 springboot 실행 이전에 github action 원격지 linux 컨테이너에서 셸스크립트로 도커 컨테이너에 설치된 MySQL에 SQL을 실행시켜야함. 청천님의 슈퍼세이브 필요
+      ddl-auto: create
```
create옵션을 운영에서 사용하면 안되는점 너무 잘 알고있으나 불가항력적인 이유로 사용했습니다.
- 사용이유 : github Action의 테스트과정에서 DB의 테이블이 없다는 이유로 build 실패 -> 테스트 실패. 
- 해결방안 : 
   1. 원격지에 접속해서 create table, update table등의 쿼리만 수동으로 실행시킨다. (X 원격지 접근불가)
   2. springBoot 시작전에 Linux ShellScript등을 통해 github Action컨테이너안의 MySQL컨테이너에 sql문을 실행시킨다. -> 시간을 너무 허비해서 저는 잠시 내려놨습니다.

heeve1 replied:
오호 그렇군요. 맞습니다. ddl-auto: create 옵션은 사용하기엔 리스크가 너무 큽니다. flyway에 대해 한번 찾아보시지요
물론 시간이 너무 소요될 것 같다면 다음에 해도 될 것 같습니다.

smh0300 committed:
feat: 회원가입 서비스 테스트 수정

heeve1 commented on .github/workflows/autotest.yml:
```diff
@@ -0,0 +1,64 @@
+
+      - name: Add Test Coverage to PR
+        id: jacoco
+        uses: madrapps/jacoco-report@v1.6.1
```
👍

heeve1 commented on src/main/java/com/Ticketease/TE/member/Member.java:
```diff
@@ -0,0 +1,38 @@
+    @Column(unique = true, nullable = false)
+    private String nickName;
+    @CreatedDate
+    private LocalDateTime createdDate;
```
updatedAt 은 딱히 사용하지 않으시려나요?

smh0300 replied:
네 계정의 최초생성시점을 스냅샷처럼 저장하는 컬럼이라서 변경이 일어나면 안되기때문에 updatedAt은 사용하지 않을예정입니다.

heeve1 commented on src/main/java/com/Ticketease/TE/member/MemberRepository.java:
```diff
@@ -0,0 +1,11 @@
+
+public interface MemberRepository extends JpaRepository<Member, Long> {
+
+    Optional<Member> findByNickName(String nickname);
```
optional 사용은 아주 좋군요. null 핸들링은 중요한 부분 중 하나입니다

heeve1 commented on src/main/java/com/Ticketease/TE/member/signUp/SignUpController.java:
```diff
@@ -0,0 +1,44 @@
+                             BindingResult bindingResult,
+                             RedirectAttributes redirectAttributes){
+        try{
+            signUpService.signUpUserValidation(SignUpRequest.toDto(request),bindingResult);
```
먼저 정답은 없습니다. controller의 역할을 어디까지 보느냐에 따라 다르게 볼 수 있을것 같은데요. 
현재 이 코드에 따르면 controller는 먼저 signUpUserValidation을 실행하고 signUpUser을 실행하는 책임까지 진것으로 보이는군요. 이 부분에 대해 좀더 고민해보시지요
controller의 역할은 어디까지로 할 것인지에 대해

smh0300 replied:
@heeve1 
다시보니 bindingResult를 사용헀을때는 Controller에서 처리하는것이 더 자연스럽네요.
그렇다면 실제 Repository와 상호작용해야하는부분을 Service에서 처리하고 그외는 Controller에서 처리하는것으로 기준을 잡으면 될까요??

heeve1 replied:
음 Controller, Service, Repository의 역할은 말씀주신 이야기로만 정의하기엔 쉽지 않아 보이는군요. 좀더 고민해보시지요!

heeve1 commented on src/main/java/com/Ticketease/TE/member/signUp/SignUpController.java:
```diff
@@ -0,0 +1,44 @@
+
+        try {
+            signUpService.signUpUser(SignUpRequest.toDto(request));
+        }catch (SignUpExceptionHandler handler){
+            redirectAttributes.addFlashAttribute(handler.getCode().toString(), handler.getMessage());
+            return "redirect:signup";
+        }
```
위의 catch(29번라인)과 사실상 같은 코드군요. try catch를 하나로 묶어도 동일한 동작을 하지 않을까 싶습니다

smh0300 replied:
네 동일한 작동을 확인했습니다. 중복을 제거하겠습니다.

heeve1 commented on src/main/java/com/Ticketease/TE/member/signUp/SignUpDto.java:
```diff
@@ -0,0 +1,17 @@
+        return new SignUpDto(nickName, password, confirmPassword);
+    }
+
+    public static Member toEntity(SignUpDto dto){
+        return Member.of(dto.nickName, dto.password);
+    }
```
오 좋습니다. 어떤 의도에서 하신 코드인지 의도를 좀더 들어보고 싶군요

smh0300 replied:
Dto를 받아서 Entity로 변환한다는 의미에서 toEntity라는 명칭을 붙였습니다.
이런방식을 사용하면 service -> Repository로 넘어가는 순간임을 명시할 수 있습니다. 또한 추후 회원에 파라미터가 추가되서 Request나 Dto를 수정했을때 발생할 수 있는 에러가 DB쪽으로 전파되지않게 막는 방파제 역할도 가능하다고 생각했습니다.

heeve1 replied:
넵 레이어 별로 별도의 모델을 정의하는 것은 의미있는 노력이라고 생각됩니다 👍

heeve1 commented on src/main/java/com/Ticketease/TE/member/signUp/SignUpService.java:
```diff
@@ -0,0 +1,36 @@
+
+    private final MemberRepository memberRepository;
+
+    public Long signUpUserValidation(SignUpDto dto, BindingResult result){
```
BindingResult를 통해서 문제가 발생한 부분은 controller에서 찾아서 클라에게 결과를 보내면 더 좋지 않을까 고민이 드는군요. 
역시 controller의 책임에 대해 조금더 고민해보시지요
(아래 result.hasErrors() 포함)

heeve1 commented on src/main/java/com/Ticketease/TE/member/signUp/SignUpService.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    public Long signUpUser(SignUpDto signUpDto){
+        Optional<Member> member = memberRepository.findByNickName(signUpDto.nickName());
+        if (member.isEmpty()) {
+            memberRepository.save(SignUpDto.toEntity(signUpDto));
+            return 0L;
+        }else{
+            throw new SignUpExceptionHandler(ExceptionCode.USER_ALREADY_EXIST, ExceptionCode.USER_ALREADY_EXIST.getDescription());
+        }
```
찾고 없다면 저장을 한다. 간단한 논리인데, 여러 요청이 동시에 발생한다면 문제는 없을까요?

smh0300 replied:
네 문제가 될것같습니다. 
- 예상되는 문제점 : findByNickName메서드가 동시에 호출되면서 member.isEmpty()를 반환한다면 같은 이름을 가진 유저가 생성될 가능성이 있습니다.
- 개선방안 : @Transactional 어노테이션을 붙여서 findByNickName과 save가 한개의 트랜젝션으로 동작할 수 있게 변경하겠습니다.

heeve1 replied:
@Tx를 붙이면 어떤 과정을 통해서 해결이 되는걸까요?

smh0300 replied:
해당내용은 다음멘토링때까지 더 자세히 알아보겠습니다

heeve1 commented on src/test/java/com/Ticketease/TE/member/signUp/SignUpServiceTest.java:
```diff
@@ -0,0 +1,95 @@
+import static org.mockito.Mockito.when;
+
+@SpringBootTest
+class SignUpServiceTest {
```
mokito를 이용한 unit test 작성은 어떨까요? @SpringBootTest 없이 단순히 java code를 검증할 수 있다면 좋겠습니다.

smh0300 replied:
넵 제안해주신 내용을 적용해서 Commit 하겠습니다.(완)

heeve1 reviewed (commented):
좋습니다. 첫 목표가 대충 돌아는 가는 서버인 만큼, 가능한 범위까지 고쳐주시면 될 것 같습니다.
다만 코멘트 별로 왜? 에 대해서는 고민을 해주시면 좋을 것 같군요!

smh0300 committed:
refactor: 중복제거, 컨트롤러 - 서비스간 역할 재분배, 테스트코드 수정

1. 중복제거 : SignUpController의 try-catch 구문 중복 제거
2. 컨트롤러 - 서비스간 역할 재분배 :
 - 서비스는 레포지토리와 접점이 있는 역할
 - 컨트롤러는 레포지토리와 접점이 없는 역할
3. 역할 재분배에 따른 테스트 코드 수정

smh0300 commented:
@kochungcheon  @heeve1 변경사항 적용했습니다!

heeve1 reviewed (approved):
고생하셨습니다~ 👍
